### ISSUE TITLE
A performance regression in v5

### ISSUE THREAD

**Issue by FransBouma** [nhibernate/nhibernate-core#1446]
Hi,

Using my raw dataaccess benchmarks (https://github.com/FransBouma/RawDataAccessBencher) which I just updated to include NHv5, latest stable, I saw the performance of NHibernate fell dramatically. v4.1 was slightly faster than EF6, but NHv5 is slower than EF6 and also consumes even more memory. 

Below is a screenshot of the set fetch of 31K rows from AdventureWorks' SalesOrderHeader, synchronously in change tracked entities. I've expanded the tree to a location which illustrates that there's something seriously wrong: with 31K rows in a *single query* it's weird the code ends up 410,000 (!) times in the method which checks whether something has to be enlisted into an ambient transaction. There's just 1 (synchronous) query. 

![nh5setfetchperf](https://user-images.githubusercontent.com/3628530/33012217-8d2dec20-cde0-11e7-946e-3875228d81f1.jpg)

You can easily profile the benchmarked fetches with a profiler, just uncomment this line: https://github.com/FransBouma/RawDataAccessBencher/blob/master/RawBencher/OriginalController.cs#L77 and follow the directions on screen. 

Hope this helps to gain some insight what to fix. Cheers.

**Comment by gliljas** [comment]
@FransBouma Thanks! We'll certainly look into this.

**Comment by hazzik** [comment]
@FransBouma can you make the same screenshot for the NH 4.1?

**Comment by fredericDelaporte** [comment]
There is also the 58.66% part which is not deployed which could be interesting to see.

**Comment by hazzik** [comment]
>I've expanded the tree to a location which illustrates that there's something seriously wrong: with 31K rows in a single query it's weird the code ends up 410,000 (!) times

Official NH's position is that the quote about should be just:

>there's something seriously wrong ... 31K rows in a single query.

My guess is that 410,000 figure is the number of entities being fetched.

**Comment by fredericDelaporte** [comment]
Thanks for the notice.

While the redundant `CheckSessionAndUpdateStatus` can certainly be changed for being way less fired (maybe with some "depth" counter in the session entry points for disabling it for sub-operations), this thing is nothing new. But it does some more checking in v5, like the transaction enlistment.

From what I read on the above screenshot, the transaction enlistment check would represent a 4% impact. It does not seem to be enough for being responsible of the "dramatic performance drop". It may be easy to check: if transaction scopes are not used, just configure the `AdoNetTransactionFactory` to be used instead of the default transaction factory. This will remove the transaction enlistment check.

Extracted from @FransBouma result files, here are the results for various things.

Test | NH v4 (2017-10-24) | NH v5 (2017-11-20) | %
------------ | -------------|-------------|------------
Change tracking fetches, set fetches (50 runs), no caching | 1.881,56ms, Enum: 2,32ms, 550.424KB | 2.354,00ms, Enum: 2,14ms, 828.909KB | +25%ms, Enum: -8%, +50%KB
Change tracking individual fetches (100 elements, 50 runs), no caching | 0,48ms, 80KB | 0,53ms, 96KB | +10%ms, +20%KB
Change tracking fetches, eager load fetches, 3-node split graph, 1000 root elements (50 runs), no caching | 273,47ms, 73.273KB | 361,18ms, 122.993KB | +32%ms, +68%KB

I hope the benchmark other conditions have not much changed between the two runs, otherwise comparing them would not make sens. (EF6 results stay similar in both runs, so I think comparing runs does make sens.)

**Comment by FransBouma** [comment]
No, it doesn't fetch the 410K entities, adventure works contains much more entities than that, and it's a connected graph, also it does loop over just the 31K rows. 

I don't know where the 410K value comes from. The source of the benchmark I used to measure it is all open source and on github, see the URL in the top post. Point it to an adventure works 2008 schema and you can profile it. The code fetches all benchmarks for a given framework first before telling you to attach a profiler, so you get warmed up results. Use a profiler which starts the exe but disabled so you can start it when you're told to do so, then start the profiler, hit enter to start the fetches, then you're told to get a snapshot and you can examine it. If you have dottrace or ants this is very easy. NHibernate loses performance all over the place though. It's not a thing where it constructs a query, runs it on the DB, obtains the reader, compiles a projection lambda/IL projector and fetches rows using it. 

The 410K value is indeed not the reason it's very slow, it was an indication that something is amiss, elsewhere along the tree there are other areas with lots of calls which shouldn't happen. 

The two resultsets quoted by @fredericDelaporte are indeed comparable, same hardware, same .NET version.

**Comment by fredericDelaporte** [comment]
The factor between fetched rows and calls can be due to the `CheckAndUpdateSessionStatus` being call on access to things like the session `PersistenceContext` property or `InitializeCollection` method.

Since `PersistenceContext` is a property, it should not do much processing. Ideally a property get should be not much more than yielding a member. Maybe should we reconsider why `CheckAndUpdateSessionStatus` is called there. If we need to keep it, check if we can instead disable it when operations like querying is ongoing. Or maybe can we replace it by just `ErrorIfClosed`. It is somewhat weird to do such checking on a property otherwise just yielding a member: what prevents the code having retrieved the property value to change the session state before using that retrieved value?

**Comment by hazzik** [comment]
I'll make a PR with a reduced amount to `ISessionImplementor.PersistentContext`.

**Comment by hazzik** [comment]
>No, it doesn't fetch the 410K entities,

Yeah, it does not. It has 250K entities referenced (either as a FK or as a fetch). The `ISessionImplementor.PersistenceContext` method was called twice in `ScheduleBatchLoadIfNeeded`.

**Comment by FransBouma** [comment]
@hazzik I have the feeling you really think it fetches more data than it should. SQL Profiler reveals the truth, it runs just *one* query: 

```sql
select salesorder0_.[SalesOrderID] as column1_46_, salesorder0_.[AccountNumber] as column2_46_, salesorder0_.[Comment] as column3_46_, salesorder0_.[CreditCardApprovalCode] as column4_46_, salesorder0_.[DueDate] as column5_46_, salesorder0_.[Freight] as column6_46_, salesorder0_.[ModifiedDate] as column7_46_, salesorder0_.[OnlineOrderFlag] as column8_46_, salesorder0_.[OrderDate] as column9_46_, salesorder0_.[PurchaseOrderNumber] as column10_46_, salesorder0_.[RevisionNumber] as column11_46_, salesorder0_.[rowguid] as column12_46_, salesorder0_.[SalesOrderNumber] as column13_46_, salesorder0_.[ShipDate] as column14_46_, salesorder0_.[Status] as column15_46_, salesorder0_.[SubTotal] as column16_46_, salesorder0_.[TaxAmt] as column17_46_, salesorder0_.[TotalDue] as column18_46_, salesorder0_.[BillToAddressID] as column19_46_, salesorder0_.[ShipToAddressID] as column20_46_, salesorder0_.[CreditCardID] as column21_46_, salesorder0_.[CurrencyRateID] as column22_46_, salesorder0_.[CustomerID] as column23_46_, salesorder0_.[SalesPersonID] as column24_46_, salesorder0_.[TerritoryID] as column25_46_, salesorder0_.[ShipMethodID] as column26_46_ from [Sales].[SalesOrderHeader] salesorder0_
```
That's it. It fetches no more data than it should: just all the rows in the table. The .NET profile results show that NHibernate does a lot of work per row which is really unnecessary. 

I recon you never ran a profiler over NHibernate?

**Comment by hazzik** [comment]
> I recon you never ran a profiler over NHibernate?

Obviously not. Why would I?

**Comment by FransBouma** [comment]
@hazzik 
> Obviously not. Why would I?

If sarcasm, why? I came to you to report a serious issue. Apparently I was wrong doing that? 
If true, you really should. Profilers reveal what's really going on, how your code really gets its results which make tests succeed. It also makes you see which parts are really inefficient.

**Comment by hazzik** [comment]
>That's it. It fetches no more data than it should: just all the rows in the table. 

I reckon you have not read my previous message or did not try to understand what I said. I said that it picked up 205K entities referenced to the entity you are trying to load. There are 8 foreign keys on the table and all of them represent many-to-one association for which NHibernate creates the proxy object.

31,465 x 8 = 251,720 (the first magical number of calls to second Hydrate). Then it filtered out unique keys and got 205,441 entities. `ScheduleBatchCallingIfNeeded` is calling `ISessionImplementor.PersistanceContext` two times, which equals to 410,882 calls (the second magical number).

---

>If sarcasm, why? 

Because it was answered in the same manner as the question asked.

>It also makes you see which parts are really inefficient.

I'm sorry to say that but at least we are not doing dirty checks on property setters that are slowing down some real-life scenarios as opposite to "look my ORM can do better than XXX" synthetic cases. 

---
Of course I ran profiler over NHibernate, but unfortunately, it does not pick up such issues when the tool is used properly.

Trying to load 31k of rows in one go is a misuse of the tool. It was always discouraged by NH developers. The common advice is to use either paging or a stateless session. 

PS: Giving me (or anyone else) here a "thumbs down" wouldn't help.

**Comment by fredericDelaporte** [comment]
I think some assumptions are done here a bit to quick, be it on technical points or on what does or think people.

> a lot of work per row which is really unnecessary

Not that sure. Yes ideally the ambient transaction checking should be done only once per "processing", like once per query. But ensuring this check is done for all cases requires to do it in quite centralized places, which may be called many time by heavy processing like the queries of the performances test. 

Removing that work could seriously undermine robustness. Of course it is still valuable to check if we can spare some of this work in a safe way, as Alexander PR #1447 is doing. So your report is any way valuable.

> It has found 205K entities referenced (either as a FK or as a fetch). 

The parenthesis here hint at the fact it is not 205k entity loaded, but 205k entity implied (referenced), part of them only as unloaded proxies.

**Comment by hazzik** [comment]
@FransBouma if you would really want to write an **honest** benchmark, then you would use stateless session for that "test'.

**Comment by gliljas** [comment]
Everyone just take a deep breath. This kind of controversy should be a thing of the past in NHibernate. The profiling result DID highlight an issue. Let's just take it constructively from there, like #1447 did.

**Comment by fredericDelaporte** [comment]
There are some other tests boiling down to "stateless", like "Non-change tracking fetches, set fetches (50 runs), no caching", but currently not implemented for NHibernate. Yes a 31k row is not a normal usage case for any ORM. Such a test it is done in my view just for pushing to the limits and eventually help at spotting some trends that may be worth checking, as illustrated by the report done here by Frans.

And although I have not checked the implementation, it looks to me other contenders in that test are running statefull. But I do not think either that the comparison between them in such an extreme case should be considered quite meaningful, since it is too far away from normal use. (And moreover, all those benchmark misses a point that I consider a premium, way more valuable than milliseconds for many business cases: the development effort saved by using the orm. For many businesses, the ability to have their product evolving fast is more important than having it sparring more cpu/memory.)

**Comment by FransBouma** [comment]
@hazzik the 'test' is to show how fast a framework can materialize rows. If it's slow it will show that. If it's fast it will show that too. You can say "that's not how it should be used", that's fine, but that doesn't mitigate the fact NHibernate, compared to others, is seriously slow in materializing objects. There's nothing really special about the query nor the data either. If I fetch 25 entities with NHibernate it still needs a lot more memory and CPU/time than other frameworks, which is unnecessary as others can do it without all that. In 1 particular query, ran just once a day it might not be noticable. But in a website where lots of users execute queries all over the place it adds up. That's why it's important. Sure the test exaggerates a flaw, that's the point: you can see it better if it's slower.  

Like I said, I'm not here to laugh at you, I'm here to notify you that *compared to the rest* things aren't as great as you might think. 

@fredericDelaporte 
> Not that sure. Yes ideally the ambient transaction checking should be done only once per "processing", like once per query. But ensuring this check is done for all cases requires to do it in quite centralized places, which may be called many time by heavy processing like the queries of the performances test.
> Removing that work could seriously undermine robustness. Of course it is still valuable to check if we can spare some of this work in a safe way, as Alexander PR #1447 is doing. So your report is any way valuable.

The ambient transaction issue was an example, if you look at the screenshot (or better, at a profile run), there are many more areas where you'd expect 1 call but there are many. 

I don't see how things can harm robustness here :) Why not look at it this way: you have a series of tasks to run when you execute a query at the session level: transform to sql, setup connections and transactions, collect data for the materializer (so just once), build the projector / materializer (or grab one from the cache), execute the sql query, pass the reader to the materializer/projector, get entity instances back. You use proxies, so the materializer has to instantiate proxy objects for related entities. That would slow things down a bit, but instantiating objects isn't slow in .NET. 

How is optimizing these tasks causing harm on robustness?

> >  It has found 205K entities referenced (either as a FK or as a fetch).
> The parenthesis here hint at the fact it is not 205k entity loaded, but 205k entity implied (referenced), part of them only as unloaded proxies.

So instantiating 205K objects is slow, you think? No it's not, sorry. I create several objects per entity instance as well. 

> There are some other tests boiling down to "stateless", like "Non-change tracking fetches, set fetches (50 runs), no caching", but currently not implemented for NHibernate. Yes a 31k row is not a normal usage case for any ORM. Such a test it is done in my view just for pushing to the limits and eventually help at spotting some trends that may be worth checking, as illustrated by the report done here by Frans.

You have change tracked and non-change tracked fetches. The non-change tracked fetches, so with a stateless session, didn't make much of a difference (as in, it was nowhere near the slowest contender) hence I didn't add it. Should I? The non-change tracked fetches are really readonly fetches so plain poco flat list fetches, and nhibernate to my knowledge didn't have such a mode. If it does, just give me a PR on the repo and I'm happy to merge it. 

Indeed, 31K rows fetching isn't 'natural' but 1) people still do that but more importantly 2) it will show if something's wrong 3) if you're fast at 31K rows, you're fast at 10-100 rows too. 

> And although I have not checked the implementation, it looks to me other contenders in that test are running statefull. 

The change tracking fetches, yes. 

> But I do not think either that the comparison between them in such an extreme case should be considered quite meaningful, since it is too far away from normal use. (And moreover, all those benchmark misses a point that I consider a premium, way more valuable than milliseconds for many business cases: the development effort saved by using the orm. For many businesses, the ability to have their product evolving fast is more important than having it sparring more cpu/memory.)

That's an orthogonal issue altogether :). This is for looking at which ORM provides the best characteristics. As more and more apps move to the cloud, memory usage + performance is crucial as the less time spend on the CPU / less memory consumed, the less money you have to pay. 

A benchmark is never a way to check whether *your app* is going to behave great. It's a way to check whether what you thought was slow/fast matches reality or not (e.g. the perception that 'Dapper' is the fastest is proven false for years but still people believe that)

**Comment by hazzik** [comment]
> If it's slow it will show that. If it's fast it will show that too.

You are comparing two cars. But you punched tyres for one of them and complaining that it did not finish.

http://hibernatingrhinos.com/products/nhprof/learn#UnboundedResultSet

**Comment by gliljas** [comment]
Let's deal with the "compared to NH4.1" scenario, first and foremost.

**Comment by FransBouma** [comment]
> You are comparing two cars. But you punched tyres for one of them and complaining that it did not finish.

Yeah whatever, keep putting your head in the sand, man, really makes the problems go away! 😄 Why not do what @gliljas does and try to make things better and invest time in fixing things instead?

**Comment by hazzik** [comment]
> PS: Giving me (or anyone else) here a "thumbs down" wouldn't help.

**Comment by hazzik** [comment]
>Why not do what @gliljas does and try to make things better and invest time in fixing things instead?

You know what? I'm trying hard for last 2 hours to run your bloody "benchmarker" which does not even compile. Don't tell people what to do, and nobody will tell you where you shall go.

**Comment by hazzik** [comment]
>You apparently are the same as Maulo

No, I'm much worse.

**Comment by hazzik** [comment]
![image](https://user-images.githubusercontent.com/144791/33067961-a787a14e-cf14-11e7-85f9-00f05a90a4f9.png)

**Comment by FransBouma** [comment]
> You know what? I'm trying hard for last 2 hours to run your bloody "benchmarker" which does not even compile. Don't tell people what to do, and nobody will tell you where you shall go.

It doesn't? Hmm. Strange, it compiles fine here, but alas, I am not a god. 

But point taken, Alexander, I won't bother you with any future issues ever again.

**Comment by hazzik** [comment]
>Instead you think I'm promoting something or trying to tell you something else is better.

That's my 10 years' experience with you and LLBLGen.

Essentially every your appearance is like this:

Frans comes in: 
 - your code sux. In LLBLGen I can fetch a million rows in a second.
 - Relax, Frans, no one does so. We know that this sux, you'll probably need to use our framework 
 _differently_ because of that.
 - No, it's really sux, you need to fix it.
 - No.

One year later:
Frans comes in:
- your code sux. In LLBLGen I can fetch a million joins in a single query.
...

**Comment by fredericDelaporte** [comment]
> That was a mistake, it seems. Well, lesson learned.

It is unfortunate things go that way. I was not commenting in the aim of criticizing/turning down, and instead I was believing being balanced. In my view I was acknowledging the goods done by your report, while attempting to explain why I am having a different view on some parts of the report.

My writings does not seem to have been understood like I was expecting. But since they are not about the main point of the issue but about side considerations, I think it is just better to let them aside.

> Don't tell people what to do, and nobody will tell you where you shall go.

I think this could be a reference to a previous part I have already quoted:

> a lot of work per row which is really unnecessary

From the ones currently maintaining the code base, this could be taken as a harsh statement, similar to "that is bad design/code, do it otherwise". I think that was not your intention, that you were just willing to point at things that we may get better. And I also think you understand there may still be some good reasons for the code to be currently as is, because each ORM has its own way of doing things, it own priorities, which may lead to some inefficiencies in some areas compared to others ORM, but also to some better things in other areas. That is why for me such sentences should more be like "it looks like there is some undue excess work that should be reduced, have a look".

And indeed, we are having a look and are doing some changes on this subject, even if we are maybe not changing it as much as you would have expected.

**Comment by FransBouma** [comment]
@hazzik Erm, no, not really. I reported an issue, you say it's nonsense, well, that's the end of it then. By all means try to ignore it and stay on this course like you're on now, [like acting like a real professional](https://github.com/FransBouma/RawDataAccessBencher/issues/36#issuecomment-346001961). 

Holy shit, imagine the poor sod who comes in here and reports a *real* issue.

**Comment by FransBouma** [comment]
@fredericDelaporte yeah sorry if I sounded a little harsh, no offense taken, you meant well and I understood your point. I also understand the codebase is complex and has a lot of history and changing it will be a large effort and likely not going to happen. I stand by my point that a lot of work along the pipeline is unnecessary _in that location_, but I understand it can come across as a general remark that it sucks, that's not how I meant it, I meant it as: "that work should probably moved elsewhere and the test/profile shows you what work are candidates for that". This might turn out to be impossible, that's ok. But at least you have tried or know these things can happen but due to its design/history etc. things are the way they are, we all have these things in our codebase. 

I was just taken aback by the storm of excuses and lameness that was poured over me, not only here but also on my own repository. I didn't spent time profiling nhibernate to give you some information to receive flak from anyone.

**Comment by hazzik** [comment]
You're really annoying, @FransBouma. The issue you've reported is being tackled and in a process of being fixed. I have put my thoughts there just not to forget what I was thinking at some point. I've become angry and shout at us like a crazy dog. I've spent good 6-7 hours trying to understand a problem and fix it. In return, you've started labelling names.

**Comment by hazzik** [comment]
On my machine combined `ISessionImplementor.PersistenceContext` is responsible "only" for 6.48%  of time. On the other hand `SessionIdLoggingContext` constructor + `Dispose` are responsible for 28.35% of time. So, this issue is _already_ resolved by #1391

**Comment by hazzik** [comment]
Duplicate of #

**Comment by hazzik** [comment]
Duplicate of #1391

**Comment by hazzik** [comment]
Here are the benchmark results for #1391 
[Benchmarks.zip](https://github.com/nhibernate/nhibernate-core/files/1495833/Benchmarks.zip)

**Comment by fredericDelaporte** [comment]
#1391 being an option to activate, I think we need to add an issue "switch that option on by default" for 6.0 milestone.

**Comment by hazzik** [comment]
I just did a memory profiling. The SessionIdLoggingContext is responsible for this enormous 800m objects allocations (in fact AsyncLocal).

**Comment by hazzik** [comment]
@fredericDelaporte I think it is fine to release 5.1 with this option being disabled by default.

**Comment by hazzik** [comment]
@fredericDelaporte I think it is fine to release 5.1 with this option being disabled by default, otherwise, I'm considering to release 6.0 instead.

**Comment by fredericDelaporte** [comment]
Ok for both options. Doing it in 5.1 will imply putting a small "possible breaking change" in the notes, but with just a config switch to change for reverting it, I think it is ok to do it.

I am also willing to test something else but have not taken the time yet: change the `using` of the `SessionIdLoggingContext` into some `using (BeginProcess())` where `BeginProcess` would belong to the abstract session, handle a depth level and do the session id stuff and the check status on first level only, then use that `BeginProcess()` instead of the using on `SessionIdLoggingContext` and instead of the calls to check status. The call to check status from properties would be an exception, where it would not do the using but just call the check status according to the current processing depth (if 0, called, otherwise not called).
It would save most of the redundant checks and set session id calls we have currently, which could greatly reduces their impact even without completely disabling the session id stuff.

**Comment by hazzik** [comment]
I might have a fix for 5.0.2. Not sure though

**Comment by fredericDelaporte** [comment]
> I am also willing to test something else but have not taken the time yet:...

Now done. 28% gain on ms (or 38% if using as base the best performing one), memory down from 830mb to 490mb. I have done a little more that what I was writing above. I will explain more in a PR. (I may not have time to do it this evening.)
